Support for regex functions using re2j-td in Presto

sopel39 opened the pull request.

sopel39 commented:
Here are our performance results:

Internal Caliper RE2J benchmarks:
https://microbenchmarks.appspot.com/runs/d2fe444e-e8a4-4713-9c06-b794775f51fb

Presto Regex benchmarks:
```
Benchmark                               (patternString)  (regexLibrary)  (sourceLength)  Mode  Cnt      Score      Error  Units
RegexpFunctionsBenchmark.benchmarkLike            .*x.*            JONI            1024  avgt   10   1816,090 ±   64,444  ns/op
RegexpFunctionsBenchmark.benchmarkLike            .*x.*            JONI           32768  avgt   10  54182,394 ± 1439,176  ns/op
RegexpFunctionsBenchmark.benchmarkLike            .*x.*            RE2J            1024  avgt   10   1407,847 ±  172,435  ns/op
RegexpFunctionsBenchmark.benchmarkLike            .*x.*            RE2J           32768  avgt   10  40120,259 ± 2558,443  ns/op
RegexpFunctionsBenchmark.benchmarkLike        .*(x|y).*            JONI            1024  avgt   10   2339,734 ±  215,585  ns/op
RegexpFunctionsBenchmark.benchmarkLike        .*(x|y).*            JONI           32768  avgt   10  70171,234 ± 3806,760  ns/op
RegexpFunctionsBenchmark.benchmarkLike        .*(x|y).*            RE2J            1024  avgt   10   5992,562 ±  286,697  ns/op
RegexpFunctionsBenchmark.benchmarkLike        .*(x|y).*            RE2J           32768  avgt   10  40932,522 ± 3103,341  ns/op
RegexpFunctionsBenchmark.benchmarkLike      longdotstar            JONI            1024  avgt   10    661,844 ±   57,854  ns/op
RegexpFunctionsBenchmark.benchmarkLike      longdotstar            JONI           32768  avgt   10  18628,748 ±  729,704  ns/op
RegexpFunctionsBenchmark.benchmarkLike      longdotstar            RE2J            1024  avgt   10   6985,772 ±  160,622  ns/op
RegexpFunctionsBenchmark.benchmarkLike      longdotstar            RE2J           32768  avgt   10  55339,934 ± 5625,846  ns/op
RegexpFunctionsBenchmark.benchmarkLike            phone            JONI            1024  avgt   10   3474,114 ±  331,038  ns/op
RegexpFunctionsBenchmark.benchmarkLike            phone            JONI           32768  avgt   10   5429,320 ±  536,348  ns/op
RegexpFunctionsBenchmark.benchmarkLike            phone            RE2J            1024  avgt   10   1492,925 ±  170,493  ns/op
RegexpFunctionsBenchmark.benchmarkLike            phone            RE2J           32768  avgt   10   1687,000 ±   79,174  ns/op
RegexpFunctionsBenchmark.benchmarkLike          literal            JONI            1024  avgt   10   1031,829 ±   61,766  ns/op
RegexpFunctionsBenchmark.benchmarkLike          literal            JONI           32768  avgt   10  31748,237 ±  899,611  ns/op
RegexpFunctionsBenchmark.benchmarkLike          literal            RE2J            1024  avgt   10   7381,229 ±  497,592  ns/op
RegexpFunctionsBenchmark.benchmarkLike          literal            RE2J           32768  avgt   10  56744,572 ± 4132,512  ns/op
```

Internal JMH benchmarks from: https://github.com/sopel39/regex-micro-benchmarks. Different patterns (see https://github.com/sopel39/regex-micro-benchmarks/blob/master/src/main/java/org/sample/MyBenchmark.java):
```
MyBenchmark.defaultRegex_pattern1    avgt   15   0,230 ± 0,005   s/op
MyBenchmark.defaultRegex_pattern2    avgt   15   0,140 ± 0,002   s/op
MyBenchmark.defaultRegex_pattern3    avgt   15   1,168 ± 0,031   s/op
MyBenchmark.joni_pattern1            avgt   15   0,203 ± 0,010   s/op
MyBenchmark.joni_pattern2            avgt   15   0,046 ± 0,005   s/op
MyBenchmark.joni_pattern3            avgt   15   0,157 ± 0,002   s/op
MyBenchmark.re2j_td_pattern1         avgt   15   0,194 ± 0,020   s/op
MyBenchmark.re2j_td_pattern2         avgt   15   0,034 ± 0,005   s/op
MyBenchmark.re2j_td_pattern3         avgt   15   0,038 ± 0,004   s/op
```

We also tested some patterns from: https://github.com/haozhun/regexp-bench/blob/master/src/main/java/com/facebook/presto/regexpbench/RegexpLibraryBenchmark.java. The data was generated using randomized generator that generates rows that match given pattern. The intent was to simulate real world Presto usage. Here are the results: https://docs.google.com/spreadsheets/d/1dErf27cpsmtTmkL3jStf5zj8m_niWqq6ukQr0ESDwcE/edit?usp=sharing.
Long row is 65000 characters. Short row is 10 characters. Data is stored in multiple ORC files. Long row tables have 10000000000 rows. Short row tables have 1538461 rows.

FYI: @martint @rschlussel @kkc00 @ilfrin

haozhun commented on presto-main/src/main/java/com/facebook/presto/operator/scalar/RegexpFunctions.java:
```diff
@@ -51,264 +43,104 @@ private RegexpFunctions()
+            throw new PrestoException(INVALID_SESSION_PROPERTY, "The setting for '" + REGEX_LIBRARY + "' is not valid. It should be either RE2J or JONI.");
+        }
+
+        int dfaStatesLimit = session.getSystemProperty(RE2J_DFA_STATES_LIMIT, Integer.class);
```
I have a question about these two session properties (state limit and retries). Can you give me some background on deciding that these two nobs need to be available for configuration?

rschlussel replied:
The states limit is used as a rough form of memory management.  RE2J has two algorithms that it can use-- DFA and NFA.  Both grow linearly with the size of the input, but DFA is substantially faster.  However, the DFA algorithm can use a ton of memory for certain inputs and pattern combinations.  Therefore we allow configuration of a state limit.  Once it hits the states limit, the re2j will fallback to using the NFA algorithm on that input.

The retry limit defines how many times re2j hits the states limit before running all of the inputs for a pattern on the nfa.  The reason for this is that whenever re2j needs to fallback to the nfa algorithm, it has to create a new dfa machine for the rest of its inputs, which is expensive if every input is going to have to fall back to nfa anyway.  However, if the inputs that hit that limit are outliers, then it's worth the cost because of the performance benefits of using the DFA on the other inputs.   You might want a higher retry limit if you are going to be running on larger datasets.

haozhun replied:
According to https://github.com/google/re2j, re2j doesn't have DFA support. I understand that it's on your roadmap to add DFA support. But I assumed you haven't. Correct me if I'm wrong.

rschlussel replied:
We forked it and implemented DFA support in our fork as well as other optmizations.  That's the artifact released as re2j-td.  (see https://github.com/Teradata/re2j)

It's not merged back to google/re2j because it operates on slices, which breaks compatibility with the current api.  (https://github.com/google/re2j/issues/17)

haozhun commented on presto-main/src/main/java/com/facebook/presto/operator/scalar/RegexpGenericPattern.java:
```diff
@@ -0,0 +1,406 @@
+
+            // Remove .* prefix and suffix. This will give performance boost when using RE2J Pattern.find() function.
+            String patternStringWithoutDotStartPrefixSuffix = patternString;
+            if (patternStringWithoutDotStartPrefixSuffix.startsWith(DOT_STAR)) {
```
Blindly getting rid of `.*` is not correct. For example, replacing `".*a"` with `""` in `"zxcba"` will yield `""`. If without `".*"`, it will be `"zxcb"`.

sopel39 replied:
Yes, that's why this is only used in `regexp_like` function (which calls `Pattern.find()`). It gives great speed improvement in DFA since it can quickly skip (fast, tight loop) through uninteresting starting characters. Generally, this corrects user suboptimal pattern when he uses `regexp_like` (see existing regex benchmarks in Presto).

haozhun replied:
This is not enough though. For example, `.*?abc` will become `?abc`. Is there any other cases that I might have missed?

sopel39 replied:
right, good catch. I will adjust for that. I think the most common case would be just .*, but it is possible that somebody would use .*? too.

haozhun commented:
We haven't decided on the exact way we are going to make the flag for controlling re2j vs joni available to `castToRegexp`. We'll get back to this later.

sopel39 commented:
I have rebased the branch and changed how system properties are propagated to `ConnectorSession`. Right now both catalog properties and system properties are accessible via `ConnectorSession.getProperty` method. This way I don't modify any SPI interface.

I was thinking about how regex properties could be propagated to `castToRegexp` function. I think that right now there are three ways to do that without major refactoring:
1. Use Java system properties (e.g. `System.getProperty`). This could work, but then regex variables would be the only ones that actually use this way of configuration. Additionally, it would not be clean to use `System.setProperty` in tests to set which library I want to verify.
2. Use configuration class (e.g. `FeatureConfig`) to register selected (regex or joni) `castToRegexp` function in `FunctionRegistry`. In such case it won't be possible to switch libraries without restarting presto (I think this is a major drawback when people will try to use Re2J version initially). It is also harder to test, because I would have to push configuration class from `AbstractTestFunctions` all the way to the `LocalQueryRunner`. This might work if there are little configuration params to be tested, but I think it will be problematic with more configuration params (e.g. pushing many configuration classes).
3. The approach that I have used currently to merge catalog properties with system properties in `FullConnectorSession`. I think that it is the easiest approach and makes testing easy (via `Session` parameter). It also does not modify SPI. I am not sure though how well it fits with your vision of such feature toggles. Maybe it's worth making RE2J library quickly available for users and then do larger refactor in this area (especially since SPI is not changed)? 

What are your opinions? @haozhun @martint @ilfrin

haozhun commented:
@sopel39 I definitely should have gotten bad to you earlier. I'm sorry.

I had a discussion with @cberner and @martint today.

Here is what I propose: We'll use config properties instead of session properties to control whether we use RE2 or JONI.

Let me explain why:

* We don't want to pass systemProperties in FullConnectorSession. This breaks an abstraction. Now, every functions, regardless of whether it's in a connector or not, will have access to all the session properties for the engine. Ideally, we should be able to pass to functions in presto-main only system properties, and functions from connectors only the specific connector's connector properties. But this is not available right now.
* Making the decision on which implementation to use inside loops can potentially introduce a highly parametric call site, which may have performance impacts. This is a potential problem for projections or filters that contains a cheap regular expression. Therefore, we would like to have two separate code instead of one code.

Here are some implementation details we propose:

* The configuration property will be added to FeaturesConfig
* RegexpType will become 2 separate types
  * Both will be registered in TypeRegistry under different names
  * RegexpGenericPattern will no longer be needed
* RegexpFunctions should be split into 2 separate implementations
  * Constructor of FunctionRegistry will take the entire FeaturesConfig instead of just isExperimentalSyntaxEnabled
    * It will choose one of JONI and RE2 implementation based on the config and register only that based on the configuration property
  * For JONI, it should be pretty straightforward
  * For RE2, it needs to take in additional parameters for `castToRegexp`
    * In order to do this, you need to implement `castToRegexp` in its own class that extends SqlScalarFunction
      * If you need an example, ArrayContains is a significantly more complex implementation of SqlScalarFunction. But you should be able to find everything you need there.
      * You can pass in the parameters through the constructor of your newly implemented `castToRegexp`
    * For the other functions, they'll be implemented in a similar way to how they are today
* In tests, push configuration class from `AbstractTestFunctions` all the way to the `LocalQueryRunner` as you described.

sopel39 commented:
@haozhun Thanks for comment. I will try to implement this as you described.

haozhun commented:
@sopel39 Thank you!

sopel39 commented:
Superseded by: https://github.com/prestodb/presto/pull/5319
